crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -112,14 +117,15 @@ pub(crate) async fn reduce<W: Worker>(
             // `reducer.should_close_batch` policy, once a batch is collected it gets
             // committed and a new batch is created with the current message.
             if reducer.should_close_batch(&batch, Some(&message)) {
-                reducer
-                    .commit(std::mem::take(&mut batch))
+                match commit_with_retry(&*reducer, Arc::new(std::mem::take(&mut batch)), &token)
                     .await
-                    .map_err(|err| {
-                        IngestionError::Reducer(format!("failed to commit batch: {err}"))
-                    })?;
-                batch = vec![message];
-                progress_update = Some(current_checkpoint_number);
+                {
+                    CommitStatus::Success => {
+                        batch = vec![message];
+                        progress_update = Some(current_checkpoint_number);
+                    }
+                    CommitStatus::Shutdown => trigger_shutdown = true,
```

Comment:
Should we also break here right away, to not try to do `let Some(message) = unprocessed.remove(&current_checkpoint_number)` again?

Reply:
No because we must report the progress of the batch.

```rust
// report progress update to executor.
 if let Some(watermark) = progress_update {
      executor_progress_sender
          .send(WorkerPoolStatus::Running((task_name.clone(), watermark)))
          .await
          .map_err(|_| IngestionError::Channel("unable to send worker pool progress updates to executor, receiver half closed".into()))?;
      progress_update = None;
}
```

Reply:
The progress update is in the outer loop. I meant breaking from the inner loop. It shouldn't affect progress update.

Reply:
Fixed: https://github.com/iotaledger/iota/pull/5820/commits/f90a75ac27131d56642d7fe76df40f77e6054459